framework/math/SpatialDiscretization/CellMappings/Lagrange/LagrangeHexMapping.cc

```diff
@@ -0,0 +1,157 @@
+#include "LagrangeHexMapping.h"
+
+#include "mesh/Cell/cell.h"
+
+#include "chi_log.h"
+
+namespace chi_math::cell_mapping
+{
+
+LagrangeHexMapping::LagrangeHexMapping(const chi_mesh::MeshContinuum& grid,
+                                       const chi_mesh::Cell& cell,
+                                       const Quadrature& volume_quadrature,
+                                       const Quadrature& surface_quadrature)
+  : LagrangeBaseMapping(grid,
+                        cell,
+                        8,
+                        MakeFaceNodeMapping(cell),
+                        volume_quadrature,
+                        surface_quadrature)
+{
+}
+
+double LagrangeHexMapping::RefShape(uint32_t i, const Vec3& qpoint) const
+{
+  ChiLogicalErrorIf(i >= 8, "Invalid shapefunction index " + std::to_string(i));
+
+  double a, b, c;
+  // clang-format off
+  if (i == 0) {a = -1.0; b=-1.0; c=-1.0;}
+  if (i == 1) {a =  1.0; b=-1.0; c=-1.0;}
+  if (i == 2) {a =  1.0; b= 1.0; c=-1.0;}
+  if (i == 3) {a = -1.0; b= 1.0; c=-1.0;}
+
+  if (i == 4) {a = -1.0; b=-1.0; c= 1.0;}
+  if (i == 5) {a =  1.0; b=-1.0; c= 1.0;}
+  if (i == 6) {a =  1.0; b= 1.0; c= 1.0;}
+  if (i == 7) {a = -1.0; b= 1.0; c= 1.0;}
+  // clang-format on
+
+  return 0.125 * (1.0 + a * qpoint.x) * (1.0 + b * qpoint.y) *
+         (1.0 + c * qpoint.z);
+}
+
+chi_mesh::Vector3 LagrangeHexMapping::RefGradShape(uint32_t i,
+                                                   const Vec3& qpoint) const
+{
+  ChiLogicalErrorIf(i >= 8, "Invalid shapefunction index " + std::to_string(i));
+
+  double a, b, c;
+  // clang-format off
+  if (i == 0) {a = -1.0; b=-1.0; c=-1.0;}
+  if (i == 1) {a =  1.0; b=-1.0; c=-1.0;}
+  if (i == 2) {a =  1.0; b= 1.0; c=-1.0;}
+  if (i == 3) {a = -1.0; b= 1.0; c=-1.0;}
+
+  if (i == 4) {a = -1.0; b=-1.0; c= 1.0;}
+  if (i == 5) {a =  1.0; b=-1.0; c= 1.0;}
+  if (i == 6) {a =  1.0; b= 1.0; c= 1.0;}
+  if (i == 7) {a = -1.0; b= 1.0; c= 1.0;}
+  // clang-format on
+
+  const double x = qpoint.x;
+  const double y = qpoint.y;
+  const double z = qpoint.z;
+
+  const double ab = a * b;
+  const double ac = a * c;
+  const double bc = b * c;
+  const double abc = a * bc;
+  return Vec3(0.125 * (a + ab * y + ac * z + abc * y * z),
+              0.125 * (b + ab * x + bc * z + abc * x * z),
+              0.125 * (c + bc * y + ac * x + abc * x * y));
```

Comment:
a for 1st component (and b for 2nd, and c for 3rd) can be factored out. I find the definition more complex because expanded. why not a*(1+b*y)(1+c*z) and similar for other components. it's also more relatable to LagrangeHexMapping::RefShape

Reply:
Done